apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt

```diff
@@ -15,416 +15,95 @@
-		val method = it["method"] as String
-		val expects = it["expects"] as? String
-		val type = parseType(it["@type"])
-
-		val form = expects?.let { OperationForm(it) }
-
-		id to Operation(id, target, type, method, form)
-	}.toMap().toMutableMap()
-}
-
-private fun parseType(type: Any?): ThingType {
-	return (type as? String)?.let { listOf(it) }
-		?: type as? ThingType
-		?: listOf()
-}
-
-private fun foldTree(jsonObject: Map<String, Any>,
-	context: Context?): Pair<MutableMap<String, Any>, MutableMap<String, Thing?>> {
-	return jsonObject
-		.filter { it.key !in listOf("@id", "@type", "@context") }
-		.entries
-		.fold(mutableMapOf<String, Any>() to mutableMapOf(), foldEntry(context))
-}
-
-typealias Embedded<T, U> = Map<T, U>
-typealias EmbeddedList<T> = List<T>
-typealias FoldedAttributes = Pair<MutableMap<String, Any>, MutableMap<String, Thing?>>
-
-private fun foldEntry(context: Context?) = { acc: FoldedAttributes, entry: Entry<String, Any> ->
-	val (attributes, things) = acc
-
-	val key = entry.key
-	val value = entry.value
-
-	when {
-		value is Embedded<*, *> -> value.parse(context, things, attributes, key)
-		value is EmbeddedList<*> && value.hasMapInstances() ->
-			value.parse(context, things, attributes, key)
-		context != null && context.isId(key) -> with(value as String) {
-			things[this] = null
-			attributes[key] = Relation(this)
-		}
-		else -> attributes[key] = value
-	}
-
-	attributes to things
-}
-
-private fun Embedded<*, *>.parse(context: Context?, things: MutableMap<String, Thing?>,
-								 attributes: MutableMap<String, Any>, key: String) {
-
-	(this as? Embedded<String, Any>)?.apply {
-		if (this.containsKey("@id")) {
-			val result = flatten(this, context)
-
-			result?.let {
-				val (thing, embeddedThings) = it
-				attributes[key] = Relation(thing.id)
-
-				things.put(thing.id, thing)
-
-				things.putAll(embeddedThings)
-			}
-		} else {
-			val (attr, embeddedThings) = foldTree(this, context)
-
-			things.putAll(embeddedThings)
-			attributes[key] = attr
-		}
-	}
-}
-
-private fun EmbeddedList<*>.parse(context: Context?, things: MutableMap<String, Thing?>,
-								  attributes: MutableMap<String, Any>, key: String) {
-
-	(this as? EmbeddedList<Embedded<String, Any>>)?.apply {
-		if (this.first().containsKey("@id")) {
-			val list = this.map { flatten(it, context)!! }
-			val mutableList = mutableListOf<Relation>()
-
-			for ((thing, embeddedThings) in list) {
-				mutableList.add(Relation(thing.id))
-
-				things.put(thing.id, thing)
-
-				things.putAll(embeddedThings)
-			}
-
-			attributes[key] = mutableList
-		} else {
-			val pairsList = this.map { foldTree(it, context) }
-
-			val foldedList = mutableListOf<Map<String, Any>>()
-
-			for ((list, embeddedThings) in pairsList) {
-				foldedList.add(list)
-
-				things.putAll(embeddedThings)
-			}
-
-			attributes[key] = foldedList
-		}
-
-	}
+/**
+ * @author Javier Gamarra
+ * @author Paulo Cruz
+ */
+object ApioConsumer {
```

Comment:
Why moving it to an object instead of functions?

Reply:
Import functions is a bit strange, at least for me, and we were using global variables to keep data in memory, for example, credentials and cache. IMHO this behavior wasn't clear to the developers, they would need to know that if they wouldn't pass the consumer would get an old value, a bit confusing.

Using object this behavior is clearer, and as we call the functions from the consumer instance then the functions will be inside the same context using the same references. 

What do you think? We could have it as classes instead, as the js consumer.
@nhpatt @victorg1991

Reply:
I would go for classes, because its the common approach

Reply:
Ok, I'll change that.